rlberry/agents/torch/sac/sac.py

```diff
@@ -0,0 +1,543 @@
+        The action is sampled from a squashed Gaussian distribution defined by the policy network.
+
+        Parameters
+        ----------
+        state: numpy.ndarray or torch.Tensor
+            The current state of the environment
+
+        Returns
+        -------
+        action torch.Tensor
+            The selected action
+        log_prob torch.Tensor
+            The log probability of the selected action
+        """
+
+        # Convert the state to a torch.Tensor if it's not already
+        state = torch.FloatTensor(state).to(self.device)
+
+        # Get the mean and log standard deviation of the action distribution from the policy network
+        action_dist = self.cont_policy(state)
+        mean, log_std = action_dist
+
+        # Compute the standard deviation and
+        # create a normal distribution with the computed mean and standard deviation
+        std = log_std.exp()
+        action_dist = torch.distributions.Normal(mean, std)
+
+        # Sample an action using the reparameterization trick
+        x_t = action_dist.rsample()
+        y_t = torch.tanh(x_t)
+
+        # Apply scaling and bias to the action
+        # and compute the log probability of the selected action
+        action = y_t * self.action_scale + self.action_bias
+        log_prob = action_dist.log_prob(x_t)
+
+        # Enforce Action Bound
+        log_prob -= torch.log(self.action_scale * (1 - y_t.pow(2)) + 1e-6)
+        log_prob = log_prob.sum(1, keepdim=True)
+        return action, log_prob
+
+    def _update(self):
+        """
+        Perform an update step for the SAC agent.
+
+        It updates the Q-networks and the policy network based on the collected
+        experiences from the replay buffer.
+        """
+
+        # Sample a batch from replay buffer
+        memory_data = self.memory.sample(self.batch_size, 1).data
+        states = (
+            torch.tensor(memory_data["states"])
+            .view(self.batch_size, -1)
+            .to(self.device)
+        )
+        next_state = (
+            torch.tensor(memory_data["next_states"])
+            .view(self.batch_size, -1)
+            .to(self.device)
+        )
+        actions = (
+            torch.tensor(memory_data["actions"])
+            .view(self.batch_size, -1)
+            .to(self.device)
+        )
+        rewards = (
+            torch.tensor(memory_data["rewards"])
+            .view(self.batch_size, -1)
+            .to(self.device)
+        )
+        dones = (
+            torch.tensor(memory_data["dones"]).view(self.batch_size, -1).to(self.device)
+        )
+
+        with torch.no_grad():
+            # Select action using the current policy
+            next_state_actions, next_state_log_pi = self._select_action(
+                next_state.detach().cpu().numpy()
+            )
+            # Compute the next state's Q-values
+            q1_next_target = self.q1_target(
+                torch.cat([next_state, next_state_actions], dim=1)
```

Comment:
Is this compatible with multidimensional shapes observations like images ? If not I suggest to use dim=-1

Reply:
I think you are right, I'll change it to `dim=-1`.